rust: update `Ref` to use the kernel's `refcount_t`.

wedsonaf opened the pull request:
Saturate the refcount instead of panicking. It also leverages existing C
code instead of reimplementing functionality.

This is the first step to deprecate `Arc` (that aborts on overflow) and `BTreeMap` (that panics on allocation failure).

Signed-off-by: Wedson Almeida Filho <wedsonaf@google.com>

nbdd0121 commented:
Aside from overflow behaviour, when should `Ref` be used but not `Arc`?

wedsonaf commented:
> Aside from overflow behaviour, when should `Ref` be used but not `Arc`?

I'm removing all usages of `Arc` from binder.

When weak references are needed, one would need to use `Arc`. But given that we currently don't have anything that needs this, I would ideally like to ban `Arc`. Eventually if the need for weak references arises, I think a saturating version would be preferable.

nbdd0121 commented:
I am actually more curious about the difference of handling ref count -- `Ref` requires the inner type to provide a `RefCount` while `Arc` supplies its own. Is the intention to create some similar to `enable_shared_from_this` so that function receivers can be "&Self"? It certainly fells "more Rust" if we instead make receiver `self: &Ref<Self>`.

wedsonaf commented:
> I am actually more curious about the difference of handling ref count -- `Ref` requires the inner type to provide a `RefCount` while `Arc` supplies its own. Is the intention to create some similar to `enable_shared_from_this` so that function receivers can be "&Self"? It certainly fells "more Rust" if we instead make receiver `self: &Ref<Self>`.

Yes, `Ref` was written so that we could "upgrade" from `&T` to `Ref<T>`.

`self: &Ref<Self>` isn't really possible AFAICT because the compiler only accepts a small set of types for `self` (which includes `Arc`). As a minor point, `&self` has the bonus advantage of avoiding the double indirection implied by `&Ref<Self>` and `&Arc<Self>`.

nbdd0121 commented:
It we are going to roll our own alloc crate then we can just make `&Ref<T>` a `Receiver`. If not, we could enable arbitrary_self_types. Manually implementing `RefCounted` needs unsafe so it's not very ideal.

Speaking of `RefCounted`, I think it should provide `inc_ref` and `dec_ref` methods instead just returning `&RefCount`; this way a C-managed wrapper can implement `RefCounted` as well, and for those defined in Rust it could just delegate to its `RefCount` instance.

ojeda commented on rust/kernel/sync/arc.rs:
```diff
@@ -49,8 +53,18 @@ unsafe impl<T: RefCounted + ?Sized + Sync + Send> Sync for Ref<T> {}
+    /// This is useful because it provides a mutable reference to `T` at its final location.
+    pub fn try_new_and_init<U: FnOnce(&mut T)>(contents: T, init: U) -> Result<Self> {
+        let mut boxed = Box::try_new(contents)?;
+        // SAFETY: We just allocated the object, no other references exist to it, so it is safe to
```
```suggestion
        // INVARIANT: The refcount is initialized to a non-zero value.
        // SAFETY: We just allocated the object, no other references exist to it, so it is safe to
```

wedsonaf replied:
Done.

ojeda commented on rust/kernel/sync/arc.rs:
```diff
@@ -30,7 +34,7 @@ use core::{
 ///
 /// The value stored in [`RefCounted::get_count`] corresponds to the number of instances of [`Ref`]
-/// that point to that instance of `T`.
+/// that point to that instance of `T`. So it is always non-zero when a [`Ref`] exists.
```
Since this is the type invariant for `Ref`, the "when a `Ref` does not exist" case cannot happen, i.e. it is always non-zero. So perhaps I would clarify that the "when" is the reason, rather than a condition:

```suggestion
/// that point to that instance of `T`. So it is always non-zero (because [`Ref`] exists).
```

wedsonaf replied:
Done.

ksquirrel commented:
Review of `e395a2f74d56`:
  - ✔️ Commit [`e395a2f`](https://github.com/Rust-for-Linux/linux/commit/e395a2f74d563a792c600442cc4e0ea50da185aa): Looks fine!

<!-- Review e395a2f74d563a792c600442cc4e0ea50da185aa -->

ojeda commented on rust/kernel/sync/arc.rs:
```diff
@@ -62,28 +76,14 @@ impl<T: RefCounted + ?Sized> Ref<T> {
-        if cur == usize::MAX {
-            panic!("Reference count overflowed");
-        }
+        // SAFETY: By the type invariant, there is necessarily a reference to the object, so it is
```
```suggestion
        // INVARIANT: C `refcount_inc` saturates, thus the refcount cannot overflow to zero.
        // SAFETY: By the type invariant, there is necessarily a reference to the object, so it is
```

wedsonaf replied:
Done.

ojeda commented on rust/kernel/sync/arc.rs:
```diff
@@ -126,22 +126,29 @@ impl<T: RefCounted + ?Sized> Clone for Ref<T> {
 
-            // Synchronises with the acquire below or with the acquire in `get_mut`.
-            if obj.get_count().count.fetch_sub(1, Ordering::Release) != 1 {
+            // SAFETY: Also by the type invariant, we are allowed to decrement the refcount. We
```
```suggestion
            // INVARIANT: If the refcount reaches zero, we are the last `Ref` and we are
            // being dropped, so nobody can observe the broken invariant.
            // SAFETY: Also by the type invariant, we are allowed to decrement the refcount. We
```

wedsonaf replied:
Done.

ksquirrel commented:
Review of `6a4673a2b906`:
  - ✔️ Commit [`6a4673a`](https://github.com/Rust-for-Linux/linux/commit/6a4673a2b9060b94f78b665ba4908919c51a019b): Looks fine!

<!-- Review 6a4673a2b9060b94f78b665ba4908919c51a019b -->

TheSven73 commented on drivers/android/process.rs:
```diff
@@ -293,22 +293,24 @@ unsafe impl Sync for Process {}
+                // SAFETY: `node_refs` is pinned behind the `Ref` reference.
+                let pinned = unsafe { Pin::new_unchecked(&process.node_refs) };
+                kernel::mutex_init!(pinned, "Process::node_refs");
+            },
```
Is the support for an initialization closure only there to support easy initialization of `Pin`ned structures? Or do you believe it's more generally useful? Just asking out of general interest.

wedsonaf replied:
See https://github.com/wedsonaf/linux/commit/61397b0c79a775f12228e47ed1ab47839d591531 -- this allows us to obviate `get_mut`, which allows us to safely claim that `Ref` is in fact pinned.

TheSven73 replied:
Ah I see ! Yes, getting rid of `get_mut()` looks very good. Not sure why you'd need a `mut` closure for initialization, though? `Arc` also provides shared references only, but we are still able to initialize the `Mutex`es inside of it?

wedsonaf replied:
See https://github.com/wedsonaf/linux/commit/e201bf1b9f5db58e5b5608d88b924d8ad5cdc14e -- allowing `init` on just `&self` (as opposed to `&mut self`) was an oversight that leads to unsafety (e.g., someone calling `init` on a synchronisation primitive that is in use).

TheSven73 replied:
Holy smoke! I was already wondering about that, but I'm coming from profound ignorance on `Pin`ning and everything around it.

nbdd0121 replied:
The best way to avoid `get_mut` is through a wrapper type like `UniqueArc` https://doc.servo.org/servo_arc/struct.UniqueArc.html

TheSven73 replied:
> allowing `init` on just `&self` (as opposed to `&mut self`) was an oversight that leads to unsafety

PS do we have an Issue to keep track of this?

wedsonaf replied:
> The best way to avoid `get_mut` is through a wrapper type like `UniqueArc` https://doc.servo.org/servo_arc/struct.UniqueArc.html

Nice, I like it.

Something for another PR though. (The reason I'm removing `get_mut` in this PR is that we don't have a `refcount_t` function with the right barriers for `get_mut` and I want this code to be simple wrappers around `refcount_t`).

> > allowing `init` on just `&self` (as opposed to `&mut self`) was an oversight that leads to unsafety
> 
> PS do we have an Issue to keep track of this?

I don't think so. Feel free to open one.

TheSven73 commented on rust/kernel/sync/arc.rs:
```diff
@@ -7,20 +7,24 @@
+use core::{cell::UnsafeCell, convert::AsRef, mem::ManuallyDrop, ops::Deref, ptr::NonNull};
+
+extern "C" {
+    fn rust_helper_refcount_set(r: *mut bindings::refcount_t, n: u32);
```
Should `n: u32` be `n: c_types::c_int` instead?
https://github.com/Rust-for-Linux/linux/blob/d8c6b9ce7fb78375460422ba15d763c22f70d0af/include/linux/refcount.h#L129-L137

wedsonaf replied:
Indeed. Thanks for catching this, I was looking at an earlier version that used `unsigned int` (https://elixir.bootlin.com/linux/v5.4.125/source/include/linux/refcount.h#L30).

TheSven73 commented on rust/helpers.c:
```diff
@@ -130,6 +130,24 @@ void rust_helper_mutex_lock(struct mutex *lock)
 }
 EXPORT_SYMBOL_GPL(rust_helper_mutex_lock);
 
+void rust_helper_refcount_set(refcount_t *r, unsigned int n)
```
Should `n` be an `int` here?
https://github.com/Rust-for-Linux/linux/blob/d8c6b9ce7fb78375460422ba15d763c22f70d0af/include/linux/refcount.h#L129-L137

TheSven73 commented on rust/kernel/sync/arc.rs:
```diff
@@ -49,8 +54,19 @@ unsafe impl<T: RefCounted + ?Sized + Sync + Send> Sync for Ref<T> {}
     pub fn try_new(contents: T) -> Result<Self> {
-        let boxed = Box::try_new(contents)?;
-        boxed.get_count().count.store(1, Ordering::Relaxed);
+        Self::try_new_and_init(contents, |_| {})
```
This feels odd. I vaguely feel that the callback should really be an `Option<Fn()>`, but I don't have strong feelings about it. If you are ok with leaving this unchanged, then I am as well.

wedsonaf replied:
The intent here was just to provide a cleaner `try_new` and `try_new_and_init`. So

```rust
  Ref::try_new(T::new());
  Ref::try_new_and_init(T::new(), |x| ...);
```

instead of

```rust
  Ref::try_new(T::new(), None);
  Ref::try_new(T::new(), Some(|x| ...));
```

The second option doesn't seem so bad when I write it out though... I think I'll merge them as you suggest.

TheSven73 replied:
Could also have a third, private, function with an `Option`, which gets called by both `try_new()` flavours? I have no strong feelings either way, whatever feels best to you.

nbdd0121 replied:
I think the current approach is okay, it is less noisy than a `Option`. I also don't think internal private function accepting `Option` is necessary, as it is pretty clear to me that `|_| {}` means a no-op.

wedsonaf replied:
Ok, I'll leave it as is for now. We can always revisit later.

TheSven73 commented on rust/kernel/sync/arc.rs:
```diff
@@ -49,8 +54,19 @@ unsafe impl<T: RefCounted + ?Sized + Sync + Send> Sync for Ref<T> {}
+        // INVARIANT: The refcount is initialised to a non-zero value.
+        // SAFETY: We just allocated the object, no other references exist to it, so it is safe to
+        // initialise the refcount to 1.
+        unsafe { rust_helper_refcount_set(boxed.get_count().count.get(), 1) };
```
I know that, in kernel C, lots of code is initializing `refcount_t` by simply zero-ing the memory, followed by a `refcount_set(n)`. Yes, it'll work.

But, is `REFCOUNT_INIT(n)` more correct? Initialization is not identical to setting an initial value, even if it happens to work for now?

wedsonaf replied:
Yes, I wanted to use `REFCOUNT_INIT` but the fact that it's an rvalue turned me off.

But since we need a helper routine anyway for `refcount_set`, we may as well replace with a `refcount_init` that just returns a value initialised with `REFCOUNT_INIT`. Then we call it in the constructor of `RefCount`.

Do you have anything better in mind?

TheSven73 replied:
Sounds good to me. In that case you won't even have to zero-init the structure using `default()`. And `#[derive(Default)]` on `RefCount` can go too.

wedsonaf replied:
> Sounds good to me. In that case you won't even have to zero-init the structure using `default()`. And `#[derive(Default)]` on `RefCount` can go too.

That's what I was planning, but we can't really trust the value in `RefCount` that is passed into `try_new_and_init` -- for example, if the caller sets the value to 0, we would be breaking the invariant. We need to initialise it again in `try_new_and_init`.

nbdd0121 replied:
Maybe we cab add the zero-refcount requirement to the safety requirement of `RefCounted`?

TheSven73 replied:
Yes, I see the problem. I'm not sure why `Ref<T>` requires its inner type to hold a `RefCount`, instead of working like `Arc<T>`? I'm sure there's a reason, I just don't know what it is.

wedsonaf replied:
> Maybe we cab add the zero-refcount requirement to the safety requirement of `RefCounted`?

I suppose we could. But given the fact that we *can* enforce this, I think it's better to just reinit the refcount in `try_new_and_init` and contain the safety requirements.

wedsonaf replied:
> Yes, I see the problem. I'm not sure why `Ref<T>` requires its inner type to hold a `RefCount`, instead of working like `Arc<T>`? I'm sure there's a reason, I just don't know what it is.

The reason for embedding the refcount is so that we can go from `&T` to `Ref<T>`, that is, we can increment the refcount on an object while having just a regular reference to it. We use it in Binder to increment the refcount on a `Process` object that implements `FileOperations`: in those functions we only get a `&Process`, using `Arc` we wouldn't have a safe path to go from `&Process` to `Arc<Process>`, which is need to do in some operations.

(We can actually build an `Arc<T>`-like abstractions from this. Perhaps we should use that instead to deprecate `Arc`.)

In my current design for red-black trees, we also only get `&T` when looking up elements stored as `Box<T>`, `Arc<T>` or `Ref<T>`, which is problematic for `Arc` when we need to take an extra reference.

nbdd0121 replied:
> I suppose we could. But given the fact that we _can_ enforce this, I think it's better to just reinit the refcount in `try_new_and_init` and contain the safety requirements.

Well, if we later change `RefCounted` to support C-manaaged pointers with `inc_ref/dec_ref` then this'll not be enforcable?

wedsonaf commented:
> It we are going to roll our own alloc crate then we can just make `&Ref<T>` a `Receiver`. If not, we could enable arbitrary_self_types. Manually implementing `RefCounted` needs unsafe so it's not very ideal.

I didn't know about the `Receiver` trait. That changes things. I'm now inclined to agree that an `Arc`-like object (backed by `refcount_t`) is preferable because it leaves things looking very much like idiomatic Rust. Then we can use something like `ArcBorrow` for the `PointerWrapper` case when we may need to increment the refcount but we don't have an `Arc` anywhere to reference.

> Speaking of `RefCounted`, I think it should provide `inc_ref` and `dec_ref` methods instead just returning `&RefCount`; this way a C-managed wrapper can implement `RefCounted` as well, and for those defined in Rust it could just delegate to its `RefCount` instance.

Yes, I think you and I briefly discussed this elsewhere. I think this PR makes it easier now that we don't really need to look into the value anymore.

ksquirrel commented:
Review of `8faf6a9bc823`:
  - ✔️ Commit [`8faf6a9`](https://github.com/Rust-for-Linux/linux/commit/8faf6a9bc82338f27dd694f8be64b113cb7c0053): Looks fine!

<!-- Review 8faf6a9bc82338f27dd694f8be64b113cb7c0053 -->

wedsonaf commented:
Updated the `Ref` implementation to be like `Arc`. (Had to land some changes before that, but they're in now.)

PTAL.

ksquirrel commented:
Review of `5743fea8ff47`:
  - ✔️ Commit [`5743fea`](https://github.com/Rust-for-Linux/linux/commit/5743fea8ff475fee4b514fc0c9869177445f01a4): Looks fine!

<!-- Review 5743fea8ff475fee4b514fc0c9869177445f01a4 -->

TheSven73 commented on rust/kernel/sync/arc.rs:
```diff
@@ -7,178 +7,216 @@
+    data: T,
 }
 
+impl<T: ?Sized> core::ops::Receiver for Ref<T> {}
```
This is so `&Ref<T>` can be used as a "self type", correct?
Perhaps add a comment line explaining what this does?

wedsonaf replied:
Yes, I'll add a comment.

ojeda commented on rust/kernel/sync/arc.rs:
```diff
@@ -7,178 +7,216 @@
-    /// Determines if two reference-counted pointers point to the same underlying instance of `T`.
-    pub fn ptr_eq(a: &Self, b: &Self) -> bool {
-        core::ptr::eq(a.ptr.as_ptr(), b.ptr.as_ptr())
+    /// Deconstructs a [`Ref`] object into a usize.
```
```suggestion
    /// Deconstructs a [`Ref`] object into a `usize`.
```

wedsonaf replied:
Done.

ojeda commented on rust/kernel/sync/arc.rs:
```diff
@@ -7,178 +7,216 @@
     }
 
-    /// Deconstructs a [`Ref`] object into a raw pointer.
+    /// Borrows a [`Ref`] instance previously deconstructued via [`Ref::into_usize`].
```
```suggestion
    /// Borrows a [`Ref`] instance previously deconstructed via [`Ref::into_usize`].
```

wedsonaf replied:
Done.

TheSven73 commented on rust/kernel/sync/arc.rs:
```diff
@@ -7,178 +7,216 @@
+    pub fn try_new_and_init<U: FnOnce(Pin<&mut T>)>(contents: T, init: U) -> Result<Self> {
+        // INVARIANT: The refcount is initialised to a non-zero value.
+        let mut inner = Box::try_new(RefInner {
+            // SAFETY: Just an FFI count that returns a `refcount_t` initilised to 1.
```
Nit: spelling?

wedsonaf replied:
Thanks! Fixing it.

ojeda reviewed (commented):
Two nits I noticed

TheSven73 commented:
I'm doing a somewhat in-depth review. Hoping to finish it today, if I can.

TheSven73 commented on rust/kernel/sync/arc.rs:
```diff
@@ -7,178 +7,216 @@
+    }
+
+    /// Returns a pinned version of a given `Ref` instance.
+    pub fn pinned(obj: Self) -> Pin<Self> {
```
Does this have a user yet? If not, would the PR be simplified by leaving out this function for now? (In my experience, abstractions without users are frequently mis-designed, i.e. it's hard to know exactly what you need, until you need it)

wedsonaf replied:
Yes, I use this in subsequent commits, for example, this [one](https://github.com/wedsonaf/linux/commit/15e0a8517db8763ffdf58d07d2a240828547bb69) where I remove `Arc` usage from the samples.

TheSven73 replied:
Sounds good, let's keep it.
I have the feeling that, since `Ref<T>` is an intrinsically pinned type, `Pin<Ref<T>>` might be redundant. But it doesn't hurt to have it, and we can always address it in a separate PR.

TheSven73 commented on rust/kernel/sync/arc.rs:
```diff
@@ -7,178 +7,216 @@
+//! 1. It is backed by the kernel's `refcount_t` type.
+//! 2. It does not support weak references, which allows it to be half the size.
+//! 3. It saturates the reference count instead of aborting when it goes over a threshold.
+//! 4. It does not provide a `get_mut` method, so the ref counted object is pinned.
```
Is the term "pinned" ambiguous here? Under some circumstances, users can safely move, or obtain a mutable reference to, a `Pin<T>`, e.g. if `T: Unpin`.

What is really meant by "pinned" here? Is it
- users cannot obtain a mutable reference, *and*
- the ref counted object will never be moved?

If so, would it make sense to document that explicitly?

wedsonaf replied:
There is no way to get a mutable reference to the `T` from a `Ref<T>`. It doesn't matter if `T: Unpin`.

Here I'm using the _pin_ meaning as in Rust, from https://doc.rust-lang.org/std/pin/index.html:

> At a high level, a `Pin<P>` ensures that the pointee of any pointer type P has a stable location in memory, meaning it cannot be moved elsewhere and its memory cannot be deallocated until it gets dropped. We say that the pointee is “pinned”. 

I suppose I can copy and paste the wording from above, but I'm not sure if this is better than saying 'pinned' -- it may give readers the impression that this has a [potentially subtle] different meaning than the regular one in Rust.

TheSven73 replied:
Yes, I agree. Especially now we've established that `try_new_and_init()` exposes a `Pin<&mut T>`, the word "pinned" is probably fully correct here.

TheSven73 commented on rust/kernel/sync/arc.rs:
```diff
@@ -7,178 +7,216 @@
+/// The reference count on an instance of [`Ref`] is always non-zero.
+/// The object pointed to by [`Ref`] is always pinned.
+pub struct Ref<T: ?Sized> {
+    ptr: NonNull<RefInner<T>>,
```
Is there a benefit to using `NonNull<T>` here instead of just `*mut T`? The [documentation](https://doc.rust-lang.org/core/ptr/struct.NonNull.html) says:
> If you’re not sure if you should use NonNull<T>, just use *mut T!

So, what is the benefit that tips the scale in favour of `NonNull<T>` ? The "must never be null" could simply be formulated as a type invariant, no?

In addition, have we thought about covariance? I find the following very hard to understand, and I wonder if we have considered its full implications?
> If your type cannot safely be covariant, you must ensure it contains some additional field to provide invariance. Often this field will be a PhantomData type like PhantomData<Cell<T>> or PhantomData<&'a mut T>.

nbdd0121 replied:
IMO `NonNull` is correct here. `Ref` is conceptually `&T` so it should be covariant. `*mut T` is invariant and is thus not correct. `NonNull` also ensures that `Option<Ref<T>>` can be stored in a single usize.

TheSven73 replied:
Thanks for the advice Gary! I posted the covariance doc quote, because I don't really understand what it means. Maybe you could provide an example that would not work (or be unsound) if we used `*mut T` instead of `NonNull<T>` ?

Edited: I guess we don't need an additional `PhantomData` field, as `Ref<T>` is covariant. Is that correct?

nbdd0121 replied:
A pretty synthetic example:
```rust
use std::ptr::NonNull;

struct Ref<T: ?Sized> {
    x: NonNull<T>,
    // x: *mut T,
}

fn take<'a>(r: Ref<&'a u32>, y: &'a u32) {}

fn give() -> Ref<&'static u32> { todo!() }

fn test() {
    let y = 1;
    take(give(), &y);
}
```
If you change x to `*mut T` the code won't compile.

With that said, we still need additional `PhantomData` field here, but it's not about variance, but rather drop check: https://doc.rust-lang.org/nightly/std/marker/struct.PhantomData.html#ownership-and-the-drop-check. As `Ref` can potentially cause `T` to be dropped, an `PhantomData` marker is still needed.

TheSven73 replied:
Thanks Gary, that's awesome! My naive view of covariance is this:
```rust
struct Dog;
struct Cat;

trait Animal {
    fn sound(&self);
}

impl Animal for Dog {
    fn sound(&self) {}
}

impl Animal for Cat {
    fn sound(&self) {}
}

let c = Arc::try_new(Cat)?;
let d = Arc::try_new(Dog)?;
let a: Arc<dyn Animal> = d; // this works ok, covariant
```
But for `Ref<T>` (our `Arc` replacement), that doesn't appear to work, what am I missing?
```rust
let c = Ref::try_new(Cat)?;
let d = Ref::try_new(Dog)?;
let a: Ref<dyn Animal> = c;
```
```
error[E0308]: mismatched types
    |
312 |         let a: Ref<dyn Animal> = c;
    |                ---------------   ^ expected trait object `dyn Animal`, found struct `Cat`
```

wedsonaf replied:
This piece is missing. Gary helped me with this on zulip, here's a link the solution: https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=b1e7c2183fd2ee4f1970b19864d6ed6b

Since it requires 3 _additional_ unstable features and we don't need it immediately (we will need it when I convert the usage of lists of `dyn DeliverToRead` in binder), I thought it would make sense to push it out to its own PR.

TheSven73 commented on rust/kernel/sync/arc.rs:
```diff
@@ -7,178 +7,216 @@
-        // to concurrently call this.
-        Some(unsafe { self.ptr.as_mut() })
+    /// This is useful because it provides a mutable reference to `T` at its final location.
+    pub fn try_new_and_init<U: FnOnce(Pin<&mut T>)>(contents: T, init: U) -> Result<Self> {
```
If I change the `FnOnce(Pin<&mut T>)` into `FnOnce(&mut T)`, then the in-kernel binder client is equally happy, i.e. the following code in `process.rs` still works:
```rust
            |process| {
                // SAFETY: `inner` is pinned behind the `Ref` reference.
                let pinned = unsafe { Pin::new_unchecked(&process.inner) };
                kernel::mutex_init!(pinned, "Process::inner");
                // SAFETY: `node_refs` is pinned behind the `Ref` reference.
                let pinned = unsafe { Pin::new_unchecked(&process.node_refs) };
                kernel::mutex_init!(pinned, "Process::node_refs");
            },
```
What does the unsafe conversion from `&mut T` to `Pin<&mut T>` buy us in this instance? I'm not saying it's necessarily redundant, I just genuinely don't know.

wedsonaf replied:
It prevents implementations from moving `process` without unsafe code (unless `T: Unpin`, but only during this call).

At the moment we don't need `process` to be mutable so it makes no different. But once e201bf1b9f5db58e5b5608d88b924d8ad5cdc14e is in, we'll need a mutable `process` to initialise mutexes.

TheSven73 replied:
Ok, that makes sense! I hadn't considered [e201bf1](https://github.com/Rust-for-Linux/linux/commit/e201bf1b9f5db58e5b5608d88b924d8ad5cdc14e). So let's keep it as `Pin<&mut T>`.

TheSven73 commented on rust/kernel/sync/arc.rs:
```diff
@@ -7,178 +7,216 @@
 
-impl<T: RefCounted + ?Sized> Drop for Ref<T> {
+impl<T: ?Sized> AsRef<T> for Ref<T> {
+    fn as_ref(&self) -> &T {
```
Does `AsRef` have an in-kernel user anywhere? If we already implement `Deref`, do we need `AsRef`?

wedsonaf replied:
Yes, it's used in linked lists at the moment for `Box` and `Arc`, and since I'm converting all `Arc` usage to `Ref`, we'll need this.

TheSven73 commented on rust/kernel/sync/arc.rs:
```diff
@@ -7,178 +7,216 @@
+    type Target = Ref<T>;
+
+    fn deref(&self) -> &Self::Target {
+        self.inner_ref.deref()
```
Any reason to prefer `self.inner.deref()` over simply `&self.inner_ref` ?

wedsonaf replied:
It's a less round-about way to express the same thing: when we use `&self.inner_ref`, we're returning a `&T` where a `&U` is expected, but given that `T: Deref<Target = U>`, the compiler coerces the value into `&U` by calling `.deref()`.

When you asked the same question on another PR, I just converted to `&X` because I saw no harm in using the compiler feature to have shorter code.

However, in another discussion I saw you trying to print a pointer and using a similar technique, but in that case it didn't work because the compiler didn't have a target type to coerce to, so it just printed the address of `&T` instead of `&U`. And now I have a reason to be careful with coercion.

TheSven73 replied:
You are absolutely correct, and well spotted! Let's keep the `.deref()`.

ojeda replied:
Coercions in general make me sad too :(

And transitive coercions tend to be evil :)

TheSven73 commented on rust/kernel/sync/arc.rs:
```diff
@@ -7,178 +7,216 @@
+            // this instance is being dropped, so the broken invariant is not observable.
+            // SAFETY: Also by the type invariant, we are allowed to decrement the refcount. We
+            // also need to limit the scope of `obj` as it cannot be accessed after the refcount is
+            // decremented.
```
> We
also need to limit the scope of `obj` as it cannot be accessed after the refcount is
decremented.

Is that statement technically correct?
Or does the `inner` reference stay valid right up until it goes out of scope at the end of the function?

wedsonaf replied:
I need to clarify that this applies only when the count is non-zero. Let's say the count went to n (> 0) after the decrement; I cannot dereference the pointer anymore because other threads/CPUs may concurrently drop it to zero and free it.

If the count reaches zero, then no other thread/CPU can possibly be using it, and the current thread/CPU is tasked with freeing it.

I'll update the comment.

wedsonaf replied:
As I was updating the code, I noticed that the comment is actually correct as I'm talking about the scope of the local reference.

I need all instances of `self.ptr.as_ref()` to be dropped before calling `Box::from_raw(self.ptr.as_ptr())` otherwise I'll have a `T` and `&T` at the same time (for the same object), which is a violation of aliasing rules.

TheSven73 replied:
Technically, there's nothing wrong with keeping the `inner` reference alive for the whole duration of the function, as long as no-one "touches it" when it's invalid. But I agree that it's better to censor `inner` starting from the point where it becomes unsound to touch it.

I am guessing that in the current code, `inner` lives slightly too long. Would it be beneficial to drop it right after it becomes unsound to touch? For example:
```rust
    fn drop(&mut self) {
        let is_zero = {
            // SAFETY: By the type invariant, there is necessarily a reference to the object.
            let inner = unsafe { self.ptr.as_ref() };
            // INVARIANT: If the refcount reaches zero, there are no other instances of `Ref`, and
            // this instance is being dropped, so the broken invariant is not observable.
            // SAFETY: Also by the type invariant, we are allowed to decrement the refcount.
            // `inner` must no longer be accessed after the refcount is decremented.
            unsafe { rust_helper_refcount_dec_and_test(inner.refcount.get()) }
        };
        if !is_zero {
            return;
        }

        // The count reached zero, we must free the memory.
        //
        // SAFETY: The pointer was initialised from the result of `Box::leak`.
        unsafe { Box::from_raw(self.ptr.as_ptr()) };
    }
```

wedsonaf replied:
> Technically, there's nothing wrong with keeping the `inner` reference alive for the whole duration of the function, as long as no-one "touches it" when it's invalid. But I agree that it's better to censor `inner` starting from the point where it becomes unsound to touch it.

Well, the argument about not touching may be true for C, not really for Rust. The mere _existence_ of incompatible references is considered UB, even if one of them is never touched. For example ([playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=017b2973a26351f92213c5d981978919)):

```rust
fn main() {
    let v = 3u32;
    let v_mut = unsafe { &mut *((&v as *const u32) as *mut u32) }; 
    println!("{}", v)
}
```

We never touch `v_mut`, but Miri flags this as UB:

```
error: Undefined Behavior: trying to reborrow for Unique at alloc1426, but parent tag <untagged> does not have an appropriate item in the borrow stack
 --> src/main.rs:3:26
  |
3 |     let v_mut = unsafe { &mut *((&v as *const u32) as *mut u32) }; 
  |                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ trying to reborrow for Unique at alloc1426, but parent tag <untagged> does not have an appropriate item in the borrow stack
  |
```


> 
> I am guessing that in the current code, `inner` lives slightly too long. Would it be beneficial to drop it right after it becomes unsound to touch? For example:
> 
> ```rust
>     fn drop(&mut self) {
>         let is_zero = {
>             // SAFETY: By the type invariant, there is necessarily a reference to the object.
>             let inner = unsafe { self.ptr.as_ref() };
>             // INVARIANT: If the refcount reaches zero, there are no other instances of `Ref`, and
>             // this instance is being dropped, so the broken invariant is not observable.
>             // SAFETY: Also by the type invariant, we are allowed to decrement the refcount.
>             // `inner` must no longer be accessed after the refcount is decremented.
>             unsafe { rust_helper_refcount_dec_and_test(inner.refcount.get()) }
>         };
>         if !is_zero {
>             return;
>         }
> 
>         // The count reached zero, we must free the memory.
>         //
>         // SAFETY: The pointer was initialised from the result of `Box::leak`.
>         unsafe { Box::from_raw(self.ptr.as_ptr()) };
>     }
> ```

You're right, it lives for too long. In fact, even your suggested version also has UB: after `rust_helper_refcount_dec_and_test` we still have `inner` (`&InnerRef<T>`) while another thread may have freed it. So I think we need to avoid it altogether while calling `rust_helper_refcount_dec_and_test`. Something like:

```rust
    fn drop(&mut self) {
        // SAFETY: By the type invariant, there is necessarily a reference to the object.
        let refcount = unsafe { self.ptr.as_ref() }.refcount.get();

        // INVARIANT: If the refcount reaches zero, there are no other instances of `Ref`, and
        // this instance is being dropped, so the broken invariant is not observable.
        // SAFETY: Also by the type invariant, we are allowed to decrement the refcount.
        let is_zero = unsafe { rust_helper_refcount_dec_and_test(refcount) };
        if is_zero {
            // The count reached zero, we must free the memory.
            //
            // SAFETY: The pointer was initialised from the result of `Box::leak`.
            unsafe { Box::from_raw(self.ptr.as_ptr()) };
        }
    }
```

Since `refcount` is a raw pointer, it's not violating any aliasing rules.

TheSven73 replied:
That's really fascinating, I had never thought about aliasing rules. Given that I'm writing unsafe Rust now, I better get up to speed quickly :)

I'm not sure if we are violating any aliasing rules simply by holding invalid references (i.e. never dereffing them). For example ([playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=fa6a11c6caa11d25a68943113cc16f84)):
```rust
use std::ptr::NonNull;
use std::boxed::Box;

fn main() {
    let ptr = NonNull::from(Box::leak(Box::new(3u32)));
    let _shared_ref = unsafe { ptr.as_ref() };
    unsafe { Box::from_raw(ptr.as_ptr()); }
    println!("I'm not touching _shared_ref");
    //println!("{}", _shared_ref);
}
```
Miri seems ok with the simple existence of the invalid `_shared_ref`, but it "has a cow" once you deref it.

That does sound compatible with the Rust aliasing rules as explained in the [`UnsafeCell` docs](https://doc.rust-lang.org/std/cell/struct.UnsafeCell.html): (emphasis mine)
> If you create a safe reference with lifetime 'a (either a &T or &mut T reference) that is accessible by safe code (for example, because you returned it), **then you must not access the data** in any way that contradicts that reference for the remainder of 'a. For example, this means that if you take the *mut T from an UnsafeCell<T> and cast it to an &T, then the data in T must remain immutable (modulo any UnsafeCell data found within T, of course) until that reference’s lifetime expires. Similarly, if you create a &mut T reference that is released to safe code, then you must not access the data within the UnsafeCell until that reference expires.

Is the goal here to prevent invalid references from ever existing? If not, could we simplify things here by not trying to limit the reference's lifetime?

But if so, then the current code is still not 100% in compliance: the `refcount: *mut bindings::refcount_t` pointer is not always valid:
```rust
        let is_zero = unsafe { rust_helper_refcount_dec_and_test(refcount) };
        // refcount invalid right here
        if is_zero {
            // refcount valid here
```
Perhaps the only way to solve this problem is by dropping the reference before `refcount_dec_and_test()` returns?
```rust
    fn drop(&mut self) {
        let is_zero = unsafe { rust_helper_refcount_dec_and_test(self.ptr.as_ref().refcount.get()) };
        if is_zero {
            unsafe { Box::from_raw(self.ptr.as_ptr()) };
        }
    }
```

wedsonaf replied:
> I'm not sure if we are violating any aliasing rules simply by holding invalid references (i.e. never dereffing them). For example ([playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=fa6a11c6caa11d25a68943113cc16f84)):
> 
> ```rust
> use std::ptr::NonNull;
> use std::boxed::Box;
> 
> fn main() {
>     let ptr = NonNull::from(Box::leak(Box::new(3u32)));
>     let _shared_ref = unsafe { ptr.as_ref() };
>     unsafe { Box::from_raw(ptr.as_ptr()); }
>     println!("I'm not touching _shared_ref");
>     //println!("{}", _shared_ref);
> }
> ```
> 
> Miri seems ok with the simple existence of the invalid `_shared_ref`, but it "has a cow" once you deref it.

It looks like the compiler is taking advantage of [non-lexical lifetimes](https://doc.rust-lang.org/edition-guide/rust-2018/ownership-and-lifetimes/non-lexical-lifetimes.html) to forget `_shared_ref` early. See this example ([playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=d35eb74a0de53974b00d150da6e12590)):

```rust
use std::ptr::NonNull;
use std::boxed::Box;

fn f(ptr: NonNull<u32>, b: &u32) {
    unsafe { Box::from_raw(ptr.as_ptr()); }
}

fn main() {
    let ptr = NonNull::from(Box::leak(Box::new(3u32)));
    let shared_ref = unsafe { ptr.as_ref() };
    f(ptr, shared_ref);
}
```

By your standards, `shared_ref` is not touched either, but Miri doesn't like it.

> Is the goal here to prevent invalid references from ever existing? If not, could we simplify things here by not trying to limit the reference's lifetime?

The goal is to prevent conflicting/dangling/invalid references from ever existing.

> But if so, then the current code is still not 100% in compliance: the `refcount: *mut bindings::refcount_t` pointer is not always valid:

It's ok to have a dangling/invalid raw pointers (`*const T` and `*mut T`) as long as you don't dereference them. It's not ok to have a dangling/invalid references (`&T` and `&mut T`).

See for example [`addr_of!`](https://doc.rust-lang.org/beta/std/ptr/macro.addr_of.html), they defined it to avoid creating an invalid reference while going from a raw pointer to another raw pointer:

> `&expr as *const _` creates a reference before casting it to a raw pointer, and that reference is subject to the same rules as all other references. This macro can create a raw pointer _without_ creating a reference first.

> Perhaps the only way to solve this problem is by dropping the reference before `refcount_dec_and_test()` returns?

I think this is fine too, but we have two unsafe operations in a single block. I want to separate them out.

TheSven73 replied:
I'm not sure what to make of the Rust aliasing rules as described in the `UnsafeCell` docs  then. I'll make abstraction of that for now :)

So yes, let's use the code we have right now, which keeps a `refcount: *mut bindings::refcount_t` around, which is invalid at certain points in the function.

Would you be able to humour me, and document briefly that a) we cannot touch the pointer after the call, and b) it's not a violation of Rust's aliasing rules?

nbdd0121 replied:
> By your standards, shared_ref is not touched either, but Miri doesn't like it.

A local reference is different from a reference in the function parameter. A function parameter is "protected" (in Miri terminology) and is marked as `dereferenceable` for LLVM, so LLVM is free to dereference it anywhere inside the function. On the other hand, with NLL, a local reference cannot be freely dereferenced by LLVM after the last usage.

TheSven73 replied:
> A local reference is different from a reference in the function parameter. A function parameter is "protected" (in Miri terminology) and is marked as `dereferenceable` for LLVM, so LLVM is free to dereference it anywhere inside the function. On the other hand, with NLL, a local reference cannot be freely dereferenced by LLVM after the last usage.

Does that mean it's valid for the shared invalid reference to exist, as long as it's not dereffed?

wedsonaf replied:
> I'm not sure what to make of the Rust aliasing rules as described in the `UnsafeCell` docs then. I'll make abstraction of that for now :)

What part of that do you think is contradictory to what I'm saying? My reading of it is: if you release a reference to safe code, then you (who hold a raw pointer) must not touch the data in incompatible ways (because safe code  may be doing it).

Further down they have the following:

> To assist with proper design, the following scenarios are explicitly declared legal for single-threaded code:
> 1. A `&T` reference can be released to safe code and there it can co-exist with other `&T` references, but not with a `&mut T`
> 1. A `&mut T` reference may be released to safe code provided neither other `&mut T` nor `&T` co-exist with it. A `&mut T` must always be unique.

Note the use of _co-exist_. They don't say anything about dereferencing.

> So yes, let's use the code we have right now, which keeps a `refcount: *mut bindings::refcount_t` around, which is invalid at certain points in the function.
> 
> Would you be able to humour me, and document briefly that a) we cannot touch the pointer after the call, and b) it's not a violation of Rust's aliasing rules?

Of course, I'll upload it shortly.

wedsonaf replied:
> Does that mean it's valid for the shared invalid reference to exist, as long as it's not dereffed?

No. Here's another example ([playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=945914d08b0c1c2793e6921913347b6f)) without a function:

```rust
use std::ptr::NonNull;
use std::boxed::Box;

fn main() {
    let ptr = NonNull::from(Box::leak(Box::new(3u32)));
    let shared_ref = unsafe { ptr.as_ref() };
    unsafe { Box::from_raw(ptr.as_ptr()); }
    println!("I'm not touching _shared_ref");
    let another_ref = shared_ref;
}
```

Here neither `shared_ref` nor `another_ref` are ever dereferenced. But the usage of `shared_ref` afterwards prevents NLL from limiting `shared_ref`'s lifetime. And miri flags UB.

nbdd0121 replied:
> Does that mean it's valid for the shared invalid reference to exist, as long as it's not dereffed?

It's possible for invalid reference to exist even in safe Rust:
```rust
let mut v = vec![1];
let x = &v[0];
v.push(2);
// now x is invalid reference
```
This is accepted by NLL. Obviously borrow checking will prevent you from accessing `x` ever again, but after borrow checking all lifetimes are erased, and for codegen `x` is just an invalid reference after push.

> ```rust
> let another_ref = shared_ref;
> ```

This is considered as an use. If you remove this then it's not UB.

TheSven73 replied:
So we have a consensus now? We'll keep a `refcount: *mut bindings::refcount_t` in scope, although it's not always valid, which nevertheless does not violate Rust's aliasing rules. And we briefly document all of this in the code.

Except @nbdd0121 if you'd like to suggest something better?

wedsonaf replied:
> > Does that mean it's valid for the shared invalid reference to exist, as long as it's not dereffed?
> 
> It's possible for invalid reference to exist even in safe Rust:
> 
> ```rust
> let mut v = vec![1];
> let x = &v[0];
> v.push(2);
> // now x is invalid reference
> ```
> 
> This is accepted by NLL. Obviously borrow checking will prevent you from accessing `x` ever again, but after borrow checking all lifetimes are erased, and for codegen `x` is just an invalid reference after push.

Right, if something is inaccessible it obviously doesn't count for the aliasing rules. If it did, something like:

```rust
fn f(p: &u32) {
    todo!()
}

fn g(q: &mut u32) {
    f(q)
}
```
Would be a violation of aliasing rules because you have a `&T` and `&mut T` at the same time. It isn't because `q: &mut u32` becomes inaccessible during the call to `f`.
 
> 
> > ```rust
> > let another_ref = shared_ref;
> > ```
> 
> This is considered as an use. If you remove this then it's not UB.

It's an use, of course, but not a dereferencing of the pointer.

nbdd0121 replied:
> So we have a consensus now? We'll keep a `refcount: *mut bindings::refcount_t` in scope, although it's not always valid, which nevertheless does not violate Rust's aliasing rules. And we briefly document all of this in the code.

The current approach looks good to me. I just want to point out that it is *okay* to have a local (but not as function parameter) dangling/invalid reference as long as it is not used, but obviously without borrow checker doing the enforcement it is *preferable* to avoid having it in scope at all.

TheSven73 reviewed (commented):
Very nice! This is a real improvement IMHO.

ksquirrel commented:
Review of `79f13db0f29d`:
  - ✔️ Commit [`79f13db`](https://github.com/Rust-for-Linux/linux/commit/79f13db0f29d1213a63fdaeb3c0282e737aeadba): Looks fine!

<!-- Review 79f13db0f29d1213a63fdaeb3c0282e737aeadba -->

wedsonaf commented:
### New version:

* Suggestions from Miguel's review
* Added comment about implementing `core::ops::Receiver`
* Fixed typo in comment in `try_new_and_init`
* Updated `drop` to avoid potential UB by breaking aliasing rules with `inner`
* Added `PhantomData` to `Ref` for drop check

bjorn3 commented on drivers/android/process.rs:
```diff
@@ -366,7 +366,7 @@ impl Process {
     }
 
-    fn set_as_manager(&self, info: Option<FlatBinderObject>, thread: &Thread) -> Result {
+    fn set_as_manager(self: &Ref<Self>, info: Option<FlatBinderObject>, thread: &Thread) -> Result {
```
Why take `self: &Ref<Self>` instead of `&self`? It isn't cloned anywhere in this function and thus should be functionally identical to `&self`, right?

bjorn3 replied:
Never mind. I see that `self.get_node` does clone it sometimes.

wedsonaf replied:
it calls `self.get_node()`, which calls `self.clone()`.

bjorn3 commented on rust/kernel/sync/arc.rs:
```diff
@@ -7,178 +7,218 @@
+    fn rust_helper_refcount_new() -> bindings::refcount_t;
+    fn rust_helper_refcount_inc(r: *mut bindings::refcount_t);
+    fn rust_helper_refcount_dec_and_test(r: *mut bindings::refcount_t) -> bool;
+}
```
Are these not exported from the `bindings` module?

TheSven73 replied:
I don't think so, they're a `static inline` defined in a header unfortunately:
https://github.com/Rust-for-Linux/linux/blob/a5a22ed136b8b032675aff92de214a7032d1ddad/include/linux/refcount.h#L265-L268
https://github.com/Rust-for-Linux/linux/blob/a5a22ed136b8b032675aff92de214a7032d1ddad/include/linux/refcount.h#L331-L334

bjorn3 replied:
I mean the `rust_helper_*` methods. Or are those invisible to bindgen?

TheSven73 replied:
I think the `rust_helper_*` methods are a work-around to be able to call header inline/static functions. bindgen does not export them. So we create a C helper function in `helpers.c` which in turn calls the inline/static header function. Now we can call the helper function in Rust.

ojeda replied:
It should work through bindgen too. We have #352 and #359.

wedsonaf replied:
I'll move along with this as it's the current state of the art (however ugly it may be). We can update it once we have more automated support.

bjorn3 commented on rust/kernel/lib.rs:
```diff
@@ -21,6 +21,7 @@
     const_panic,
     const_raw_ptr_deref,
     const_unreachable_unchecked,
+    receiver_trait,
```
This feature is not meant to eventually be stabilized. It doesn't have a tracking issue and `Receiver` is marked as `#[doc(hidden)]`.

nbdd0121 replied:
Given that we are going to have a in-tree copy of `alloc` crate I think it's okay to use it for now. We'll need `DispatchFromDyn` anyway which is also hidden. I think eventually we should push for a stable way in Rust to define custom smart pointers that are as usable as `Arc`.

bjorn3 replied:
`Receiver` could be stabilized if wanted. For `DispatchFromDyn` this is not as easy as by definition it relies on an unstable implementation detail. An incorrect implementation of it can cause an ICE or miscompilation I think. Basically for `Foo` to have a valid `DispatchFromDyn` impl, `Foo<dyn Trait>` must consist of only a single fat pointer with the data part being identical to the thin pointer in `Foo<ConcreteType>`. Rustc assumes that `Foo<dyn Trait>` is passed in two registers. The second register contains the vtable and is used to get the method function pointer. The first register is then passed as sole argument to the method. Turning `Foo<ConcreteType>` into `Foo<dyn Trait>` also depends on another unstable feature called `CoerceUnsized`, which itself depends on `Unsize`.

wedsonaf replied:
To add some more context to what Gary already said: we had a technical meeting on the `alloc` crate last week. The consensus was that we'll have a "fork" of `alloc` in the kernel tree; changes directly to it are discouraged but in the event that they are needed, we will work with the Rust community to upstream them so in steady-state our fork and upstream should be the same. 

A replacement for `Arc` was explicitly discussed, and @joshtriplett mentioned that it may be reasonable for Rust to allow `Arc`-like implementations that are not so tied to the compiler, and that we should explore this as a way to minimise changes to our fork of `alloc`. While this isn't a reality, and once our fork is committed, we can move this `Ref` implementation to our fork of `alloc` if we want to minimise the risk of other components accidentally relying on unstable features required by `Ref`.

ksquirrel commented:
Review of `6ba4d643feb8`:
  - ✔️ Commit [`6ba4d64`](https://github.com/Rust-for-Linux/linux/commit/6ba4d643feb8d138518512da9598893076ed8808): Looks fine!

<!-- Review 6ba4d643feb8d138518512da9598893076ed8808 -->

wedsonaf commented:
Updated comment. PTAL.
